[DataLoader] Make `shuffle`  DataPipe sensitive to unbatch_level argument

VitalyFedyunin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60840 [WIP] Example of DataPipes and DataFrames integration
* #60680 [DataLoader] Add notebook with DataPipes API example
* #60679 [DataLoader] Add Example Only `fork` DataPipe
* #60678 [DataLoader] Decorate ZipDataPipe as `zip`
* #60675 [DataLoader] Add simple `groupby` DataPipe
* #60672 [DataLoader] Make `batch` DataPipe sensitive to unbatch_level argument
* **#60671 [DataLoader] Make `shuffle`  DataPipe sensitive to unbatch_level argument**
* #60670 [DataLoader] Rename transform DataPipe to legacy_transform

Differential Revision: [D29461083](https://our.internmc.facebook.com/intern/diff/D29461083)

VitalyFedyunin committed:
[DataLoader] Make `shuffle`  DataPipe sensitive to unbatch_level argument

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[DataLoader] Make `shuffle`  DataPipe sensitive to unbatch_level argument"




[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/iter/combinatorics.py:
```diff
@@ -74,19 +76,29 @@ class ShuffleIterDataPipe(IterDataPipe[T_co]):
+        return val
 
     def __iter__(self) -> Iterator[T_co]:
+        # TODO: Buffer is global, should be per __iter__ !!!
```
Is this TODO aiming to make shuffle shardable?

VitalyFedyunin replied:
It is just standard behaviour of iterators to be independent from each other

ejguan reviewed (approved):
LGTM, one question as below.

VitalyFedyunin committed:
Update on "[DataLoader] Make `shuffle`  DataPipe sensitive to unbatch_level argument"




[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461083).

VitalyFedyunin committed:
Update on "[DataLoader] Make `shuffle`  DataPipe sensitive to unbatch_level argument"


Differential Revision: [D29461083](https://our.internmc.facebook.com/intern/diff/D29461083)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461083).

VitalyFedyunin committed:
Update on "[DataLoader] Make `shuffle`  DataPipe sensitive to unbatch_level argument"


Differential Revision: [D29461083](https://our.internmc.facebook.com/intern/diff/D29461083)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29461083).
